go/libraries/doltcore/sqle/database.go

```diff
@@ -44,62 +47,129 @@ const (
 	single  batchMode = false
 )
 
+type tableCache struct {
+	mu *sync.Mutex
+	tables map[string]map[*doltdb.RootValue]map[string]sql.Table
+}
+
+func (tc *tableCache) Get(dbName string, tableName string, root *doltdb.RootValue) (sql.Table, bool) {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+
+	dbTables, ok := tc.tables[dbName]
+
+	if !ok {
+		return nil, false
+	}
+
+	tablesForRoot, ok := dbTables[root]
+
+	if !ok {
+		return nil, false
+	}
+
+	tbl, ok := tablesForRoot[tableName]
+
+	return tbl, ok
+}
+
+func (tc *tableCache) Put(dbName string, tableName string, root *doltdb.RootValue, tbl sql.Table) {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+
+	dbTables, ok := tc.tables[dbName]
+
+	if !ok {
+		dbTables = make(map[*doltdb.RootValue]map[string]sql.Table)
+		tc.tables[dbName] = dbTables
+	}
+
+	tablesForRoot, ok := dbTables[root]
+
+	if !ok {
+		tablesForRoot = make(map[string]sql.Table)
+		dbTables[root] = tablesForRoot
+	}
+
+	tablesForRoot[tableName] = tbl
+}
+
+func (tc *tableCache) AllForRoot(dbName string, root *doltdb.RootValue) (map[string]sql.Table, bool) {
+	tc.mu.Lock()
+	defer tc.mu.Unlock()
+
+	dbTables, ok := tc.tables[dbName]
+
+	if !ok {
+		return nil, false
+	}
+
+	tablesForRoot, ok := dbTables[root]
+	return tablesForRoot, ok
+}
+
+var tblCache = &tableCache{&sync.Mutex{}, make(map[string]map[*doltdb.RootValue]map[string]sql.Table)}
+
 // Database implements sql.Database for a dolt DB.
 type Database struct {
 	name      string
-	root      *doltdb.RootValue
+	defRoot   *doltdb.RootValue
 	ddb       *doltdb.DoltDB
 	rsr       env.RepoStateReader
 	batchMode batchMode
-	tables    map[*doltdb.RootValue]map[string]sql.Table
 }
 
-var _ sql.Database = (*Database)(nil)
-var _ sql.VersionedDatabase = (*Database)(nil)
-var _ sql.TableDropper = (*Database)(nil)
-var _ sql.TableCreator = (*Database)(nil)
-var _ sql.TableRenamer = (*Database)(nil)
+var _ sql.Database = Database{}
+var _ sql.VersionedDatabase = Database{}
+var _ sql.TableDropper = Database{}
+var _ sql.TableCreator = Database{}
+var _ sql.TableRenamer = Database{}
 
 // NewDatabase returns a new dolt database to use in queries.
-func NewDatabase(name string, root *doltdb.RootValue, ddb *doltdb.DoltDB, rsr env.RepoStateReader) *Database {
-	return &Database{
+func NewDatabase(name string, defRoot *doltdb.RootValue, ddb *doltdb.DoltDB, rsr env.RepoStateReader) Database {
```

Comment:
One of the larger changes is that I changed all the receivers for Database to be value types.  The Database object is immutable.

Reply:
Why? Doesn't this result in a struct copy on every method call?

Reply:
It does. But I don't really see this as a performance issue.  This seems to be the way go prefers to do immutability unless the struct gets large.  Taking this conversation to gchat to get people's thoughts.